### PR TITLE
chore(settings): dataroot and wwwroot now in settings.php

### LINKED ISSUE #10355: Require wwwroot, dataroot, and cacheroot in settings.php
This would eliminate a ton of code, race conditions, and general headaches like #10351.

Probably would need to (get to) drop the web installer.

### PR DESCRIPTION
**Pull request by mrclay**
The installer now sets `dataroot` and `wwwroot` in settings.php.
`$CONFIG->dataroot` is required for operation.
If `$CONFIG->wwwroot` is missing, Symfony Request will compute it.
Creating a 2nd ElggSite can no longer be done.
The sites_entity table URL no longer stores the URL to avoid confusion.

Fixes #10355

BREAKING CHANGES:
See `docs/guides/upgrading.rst` for details.

- [x] upgrade worked fine

### CODE REVIEW

**Review by hypeJunction — COMMENTED** [review]
:clap:

**Comment by jdalsem** [comment]
Did you test upgrading a 2.3 to master installation? Maybe the docs should clarify editing the settings.php should be done before upgrading...

**Comment by mrclay** [comment]
Have not tested a 2.3 upgrade, but without dataroot it will echo a message and exit right after reading config. I'll make docs clearer and test a 2.3

File: engine/classes/ElggInstaller.php [hunk 1]
```diff
@@ -1207,6 +1200,44 @@ protected function validateDatabaseVars($submissionVars, $formVars) {
...
+		} else {
+			if (strpos($submissionVars['dataroot'], '/') !== 0) {
+				$msg = _elgg_services()->translator->translate('install:error:relative_path', [$submissionVars['dataroot']]);
+				register_error($msg);
+				return false;
+			}
+		}
+
+		// check that data root exists
+		if (!file_exists($submissionVars['dataroot'])) {
```
Comment (jeabakker) [inline engine/classes/ElggInstaller.php]: wouldn't it be better to check `is_dir()`?

File: engine/tests/phpunit/ElggSiteTest.php [hunk 2]
```diff
@@ -16,7 +16,7 @@ public function testGetNoreplyEmailAddress() {
 		$site = new \ElggSite();
 		$site->url = 'https://example.com/';
 		
-		$this->assertEquals('noreply@example.com', $site->getEmailAddress());
+		$this->assertEquals('noreply@localhost', $site->getEmailAddress());
```
Comment (jeabakker) [inline engine/tests/phpunit/ElggSiteTest.php]: maybe explain why this isn't the domain set 2 lines higher

File: actions/admin/site/update_advanced.php [hunk 3]
```diff
@@ -17,28 +17,6 @@
 	throw new InstallationException("Passing a non-ElggSite to an ElggSite constructor!");
 }
 
-$site->url = rtrim(get_input('wwwroot', '', false), '/') . '/';
-
-elgg_save_config('path', sanitise_filepath(get_input('path', '', false)));
```
Comment (jeabakker) [inline actions/admin/site/update_advanced.php]: `path` is no longer save, where does it come from if i ask `elgg_get_config('path')`?

I see `elgg_get_root_path()` already is based on different logic, was this documented?

**Comment by mrclay** [comment]
Upgrade worked.

**Pull request merged**